nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/smtp/SmtpConsumer.java

```diff
@@ -0,0 +1,161 @@
+ * provided by {@link SMTPServer} and NiFi polling scheduler mechanism.
+ */
+public class SmtpConsumer implements MessageHandler {
+
+    private String from = null;
+    private final List<String> recipientList = new ArrayList<>();
+    private final MessageContext context;
+    private final ProcessSessionFactory sessionFactory;
+    private final int port;
+    private final int maxMessageSize;
+    private final ComponentLog log;
+    private final String host;
+
+    public SmtpConsumer(
+            final MessageContext context,
+            final ProcessSessionFactory sessionFactory,
+            final int port,
+            final String host,
+            final ComponentLog log,
+            final int maxMessageSize
+    ) {
+        this.context = context;
+        this.sessionFactory = sessionFactory;
+        this.port = port;
+        if (host == null || host.trim().isEmpty()) {
+            this.host = context.getSMTPServer().getHostName();
+        } else {
+            this.host = host;
+        }
+        this.log = log;
+        this.maxMessageSize = maxMessageSize;
+    }
+
+    String getFrom() {
+        return from;
+    }
+
+    List<String> getRecipients() {
+        return Collections.unmodifiableList(recipientList);
+    }
+
+    @Override
+    public void data(final InputStream data) throws RejectException, TooMuchDataException, IOException {
+        final ProcessSession processSession = sessionFactory.createSession();
+        final StopWatch watch = new StopWatch();
+        watch.start();
+        try {
+            FlowFile flowFile = processSession.create();
+            final AtomicBoolean limitExceeded = new AtomicBoolean(false);
+            flowFile = processSession.write(flowFile, (OutputStream out) -> {
+                final LimitingInputStream lis = new LimitingInputStream(data, maxMessageSize);
+                IOUtils.copy(lis, out);
+                if (lis.hasReachedLimit()) {
+                    limitExceeded.set(true);
+                }
+            });
+            if (limitExceeded.get()) {
+                throw new TooMuchDataException("Maximum message size limit reached - client must send smaller messages");
+            }
+            flowFile = processSession.putAllAttributes(flowFile, extractMessageAttributes());
+            watch.stop();
+            processSession.getProvenanceReporter().receive(flowFile, "smtp://" + host + ":" + port + "/", watch.getDuration(TimeUnit.MILLISECONDS));
+            processSession.transfer(flowFile, ListenSMTP.REL_SUCCESS);
+            processSession.commit();
+        } catch (FlowFileAccessException | IllegalStateException | RejectException | IOException ex) {
+            log.error("Unable to fully process input due to " + ex.getMessage(), ex);
+            throw ex;
+        } finally {
+            processSession.rollback(); //make sure this happens no matter what - is safe
```

Comment:
@joewitt fascinating refactor. 

Would you mind explaining why the rollback is used and why it doesn't affect the commit above?